File: src/swift/FluxConstraint.swift
import Foundation

#if canImport(XCTest)
import XCTest
#endif

// MARK: - Severity Constants

public enum FluxSeverity {
    public static let PASS = 0
    public static let CAUTION = 1
    public static let WARNING = 2
    public static let CRITICAL = 3
}

// MARK: - FluxResult Structure

public struct FluxResult {
    public let errorMask: UInt8       // Bitmask of violated constraints (0-7)
    public let severity: Int          // Highest severity level encountered
    public let violatedLo: Int8       // Lowest violated threshold
    public let violatedHi: Int8       // Highest violated threshold
    public let checkedValue: Int8     // The clamped input value that was checked

    public init(errorMask: UInt8 = 0, severity: Int = FluxSeverity.PASS,
                violatedLo: Int8 = 0, violatedHi: Int8 = 0, checkedValue: Int8 = 0) {
        self.errorMask = errorMask
        self.severity = severity
        self.violatedLo = violatedLo
        self.violatedHi = violatedHi
        self.checkedValue = checkedValue
    }
}

// MARK: - FluxConstraint Structure

public struct FluxConstraintItem {
    public let loThreshold: Int8      // Lower threshold
    public let hiThreshold: Int8      // Upper threshold
    public let severity: Int          // Severity level for this constraint

    public init(loThreshold: Int8, hiThreshold: Int8, severity: Int) {
        self.loThreshold = loThreshold
        self.hiThreshold = hiThreshold
        self.severity = severity
    }
}

// MARK: - BatchStats Structure

public struct BatchStats {
    public let totalChecks: Int
    public let violationCount: Int
    public let maxSeverity: Int
    public let avgSeverity: Double
    public let processingTime: TimeInterval

    public init(totalChecks: Int, violationCount: Int, maxSeverity: Int,
                avgSeverity: Double, processingTime: TimeInterval) {
        self.totalChecks = totalChecks
        self.violationCount = violationCount
        self.maxSeverity = maxSeverity
        self.avgSeverity = avgSeverity
        self.processingTime = processingTime
    }
}

// MARK: - FluxConstraint Main Class

public class FluxConstraint {
    private var constraints: [FluxConstraintItem] = []
    private let maxConstraints = 8
    public let sensorName: String

    public init(sensorName: String = "unknown") {
        self.sensorName = sensorName
    }

    // MARK: - INT8 Saturation

    private func saturateINT8(_ value: Int) -> Int8 {
        return Int8(max(-127, min(127, value)))
    }

    // MARK: - Constraint Management

    public func addConstraint(loThreshold: Int8, hiThreshold: Int8, severity: Int) throws {
        guard constraints.count < maxConstraints else {
            throw FluxError.tooManyConstraints("Maximum 8 constraints allowed")
        }

        guard loThreshold <= hiThreshold else {
            throw FluxError.invalidThresholds("Low threshold cannot be greater than high threshold")
        }

        let constraint = FluxConstraintItem(loThreshold: loThreshold, hiThreshold: hiThreshold, severity: severity)
        constraints.append(constraint)
    }

    // MARK: - Single Value Checking

    public func check(_ value: Int) -> FluxResult {
        let clampedValue = saturateINT8(value)

        var errorMask: UInt8 = 0
        var maxSeverity = FluxSeverity.PASS
        var violatedLo: Int8 = 127
        var violatedHi: Int8 = -127

        for (index, constraint) in constraints.enumerated() {
            if clampedValue < constraint.loThreshold || clampedValue > constraint.hiThreshold {
                // Constraint violated
                errorMask |= (1 << UInt8(index))

                if constraint.severity > maxSeverity {
                    maxSeverity = constraint.severity
                }

                if constraint.loThreshold < violatedLo {
                    violatedLo = constraint.loThreshold
                }

                if constraint.hiThreshold > violatedHi {
                    violatedHi = constraint.hiThreshold
                }
            }
        }

        // Reset violated thresholds if no violations
        if errorMask == 0 {
            violatedLo = 0
            violatedHi = 0
        }

        return FluxResult(errorMask: errorMask, severity: maxSeverity,
                         violatedLo: violatedLo, violatedHi: violatedHi,
                         checkedValue: clampedValue)
    }

    // MARK: - Batch Processing

    public func checkBatch(_ values: [Int]) -> (results: [FluxResult], stats: BatchStats) {
        let startTime = CFAbsoluteTimeGetCurrent()

        let results = values.map { check($0) }
        let violationCount = results.filter { $0.errorMask != 0 }.count
        let maxSeverity = results.map { $0.severity }.max() ?? FluxSeverity.PASS
        let avgSeverity = results.isEmpty ? 0.0 : Double(results.map { $0.severity }.reduce(0, +)) / Double(results.count)

        let processingTime = CFAbsoluteTimeGetCurrent() - startTime

        let stats = BatchStats(totalChecks: values.count, violationCount: violationCount,
                              maxSeverity: maxSeverity, avgSeverity: avgSeverity,
                              processingTime: processingTime)

        return (results, stats)
    }

    // MARK: - Industry Presets

    public static let industryPresets: [String: [FluxConstraintItem]] = [
        "aviation": [
            FluxConstraintItem(loThreshold: -100, hiThreshold: 100, severity: FluxSeverity.PASS),
            FluxConstraintItem(loThreshold: -80, hiThreshold: 80, severity: FluxSeverity.CAUTION),
            FluxConstraintItem(loThreshold: -60, hiThreshold: 60, severity: FluxSeverity.WARNING),
            FluxConstraintItem(loThreshold: -40, hiThreshold: 40, severity: FluxSeverity.CRITICAL)
        ],
        "medical": [
            FluxConstraintItem(loThreshold: -50, hiThreshold: 50, severity: FluxSeverity.PASS),
            FluxConstraintItem(loThreshold: -40, hiThreshold: 40, severity: FluxSeverity.CAUTION),
            FluxConstraintItem(loThreshold: -30, hiThreshold: 30, severity: FluxSeverity.WARNING),
            FluxConstraintItem(loThreshold: -20, hiThreshold: 20, severity: FluxSeverity.CRITICAL)
        ],
        "maritime": [
            FluxConstraintItem(loThreshold: -120, hiThreshold: 120, severity: FluxSeverity.PASS),
            FluxConstraintItem(loThreshold: -90, hiThreshold: 90, severity: FluxSeverity.CAUTION),
            FluxConstraintItem(loThreshold: -60, hiThreshold: 60, severity: FluxSeverity.WARNING),
            FluxConstraintItem(loThreshold: -30, hiThreshold: 30, severity: FluxSeverity.CRITICAL)
        ],
        "automotive": [
            FluxConstraintItem(loThreshold: -110, hiThreshold: 110, severity: FluxSeverity.PASS),
            FluxConstraintItem(loThreshold: -85, hiThreshold: 85, severity: FluxSeverity.CAUTION),
            FluxConstraintItem(loThreshold: -55, hiThreshold: 55, severity: FluxSeverity.WARNING),
            FluxConstraintItem(loThreshold: -25, hiThreshold: 25, severity: FluxSeverity.CRITICAL)
        ],
        "energy": [
            FluxConstraintItem(loThreshold: -127, hiThreshold: 127, severity: FluxSeverity.PASS),
            FluxConstraintItem(loThreshold: -100, hiThreshold: 100, severity: FluxSeverity.CAUTION),
            FluxConstraintItem(loThreshold: -75, hiThreshold: 75, severity: FluxSeverity.WARNING),
            FluxConstraintItem(loThreshold: -50, hiThreshold: 50, severity: FluxSeverity.CRITICAL)
        ]
    ]

    public func fromPreset(_ presetName: String) throws {
        guard let presetConstraints = FluxConstraint.industryPresets[presetName] else {
            throw FluxError.unknownPreset("Unknown preset: \(presetName)")
        }

        constraints.removeAll()

        for constraint in presetConstraints {
            try addConstraint(loThreshold: constraint.loThreshold,
                             hiThreshold: constraint.hiThreshold,
                             severity: constraint.severity)
        }
    }

    // MARK: - Benchmarking

    public func benchmark(iterations: Int = 10000) throws -> Double {
        guard iterations > 0 else {
            throw FluxError.invalidParameter("Iterations must be positive")
        }

        // Prepare test data
        let testValues = (0..<1000).map { ($0 % 255) - 127 }

        let startTime = CFAbsoluteTimeGetCurrent()
        var checksPerformed = 0

        for _ in 0..<iterations {
            for value in testValues {
                _ = check(value)
                checksPerformed += 1
            }
        }

        let duration = CFAbsoluteTimeGetCurrent() - startTime
        return Double(checksPerformed) / duration
    }
}

// MARK: - Error Types

public enum FluxError: Error, LocalizedError {
    case tooManyConstraints(String)
    case invalidThresholds(String)
    case unknownPreset(String)
    case invalidParameter(String)

    public var errorDescription: String? {
        switch self {
        case .tooManyConstraints(let message),
             .invalidThresholds(let message),
             .unknownPreset(let message),
             .invalidParameter(let message):
            return message
        }
    }
}

// MARK: - Unit Tests

#if canImport(XCTest)

public class FluxConstraintTests: XCTestCase {

    func testSaturateINT8() {
        let fc = FluxConstraint(sensorName: "test")

        XCTAssertEqual(fc.saturateINT8(0), 0)
        XCTAssertEqual(fc.saturateINT8(127), 127)
        XCTAssertEqual(fc.saturateINT8(128), 127)
        XCTAssertEqual(fc.saturateINT8(-127), -127)
        XCTAssertEqual(fc.saturateINT8(-128), -127)
        XCTAssertEqual(fc.saturateINT8(1000), 127)
        XCTAssertEqual(fc.saturateINT8(-1000), -127)
    }

    func testAddConstraint() {
        let fc = FluxConstraint(sensorName: "test")

        // Test normal addition
        XCTAssertNoThrow(try fc.addConstraint(loThreshold: -50, hiThreshold: 50, severity: FluxSeverity.WARNING))

        // Test invalid threshold order
        XCTAssertThrowsError(try fc.addConstraint(loThreshold: 50, hiThreshold: -50, severity: FluxSeverity.WARNING))

        // Test maximum constraints
        for _ in 1..<8 {
            try! fc.addConstraint(loThreshold: -10, hiThreshold: 10, severity: FluxSeverity.PASS)
        }

        XCTAssertThrowsError(try fc.addConstraint(loThreshold: -5, hiThreshold: 5, severity: FluxSeverity.PASS))
    }

    func testBasicCheck() {
        let fc = FluxConstraint(sensorName: "test")
        try! fc.addConstraint(loThreshold: -50, hiThreshold: 50, severity: FluxSeverity.WARNING)

        // Test within bounds
        let result1 = fc.check(25)
        XCTAssertEqual(result1.errorMask, 0)

        // Test out of bounds
        let result2 = fc.check(75)
        XCTAssertNotEqual(result2.errorMask, 0)
        XCTAssertEqual(result2.severity, FluxSeverity.WARNING)
    }

    func testMultipleConstraints() {
        let fc = FluxConstraint(sensorName: "test")
        try! fc.addConstraint(loThreshold: -100, hiThreshold: 100, severity: FluxSeverity.PASS)
        try! fc.addConstraint(loThreshold: -50, hiThreshold: 50, severity: FluxSeverity.CAUTION)
        try! fc.addConstraint(loThreshold: -25, hiThreshold: 25, severity: FluxSeverity.WARNING)
        try! fc.addConstraint(loThreshold: -10, hiThreshold: 10, severity: FluxSeverity.CRITICAL)

        let result = fc.check(75)

        // Should violate constraints 1, 2, and 3 (bits 1, 2, 3 set)
        let expectedMask: UInt8 = 0b00001110
        XCTAssertEqual(result.errorMask, expectedMask)
        XCTAssertEqual(result.severity, FluxSeverity.CRITICAL)
    }

    func testCheckBatch() {
        let fc = FluxConstraint(sensorName: "test")
        try! fc.addConstraint(loThreshold: -50, hiThreshold: 50, severity: FluxSeverity.WARNING)

        let values = [-25, 0, 25, 75, -75]
        let (results, stats) = fc.checkBatch(values)

        XCTAssertEqual(results.count, values.count)
        XCTAssertEqual(stats.totalChecks, values.count)
        XCTAssertEqual(stats.violationCount, 2)
    }

    func testFromPreset() {
        let fc = FluxConstraint(sensorName: "test")

        let presets = ["aviation", "medical", "maritime", "automotive", "energy"]

        for preset in presets {
            XCTAssertNoThrow(try fc.fromPreset(preset))
            XCTAssertGreaterThan(fc.constraints.count, 0)
        }

        // Test invalid preset
        XCTAssertThrowsError(try fc.fromPreset("invalid"))
    }

    func testBenchmark() {
        let fc = FluxConstraint(sensorName: "test")
        try! fc.fromPreset("aviation")

        XCTAssertNoThrow(try fc.benchmark(iterations: 100))

        let rate = try! fc.benchmark(iterations: 100)
        XCTAssertGreaterThan(rate, 0)

        // Test invalid iterations
        XCTAssertThrowsError(try fc.benchmark(iterations: 0))
    }

    func testSaturationBehavior() {
        let fc = FluxConstraint(sensorName: "test")
        try! fc.addConstraint(loThreshold: -50, hiThreshold: 50, severity: FluxSeverity.WARNING)

        // Test extreme values get saturated
        let result1 = fc.check(1000)
        XCTAssertEqual(result1.checkedValue, 127)

        let result2 = fc.check(-1000)
        XCTAssertEqual(result2.checkedValue, -127)
    }

    func testViolatedThresholds() {
        let fc = FluxConstraint(sensorName: "test")
        try! fc.addConstraint(loThreshold: -30, hiThreshold: 30, severity: FluxSeverity.WARNING)
        try! fc.addConstraint(loThreshold: -60, hiThreshold: 60, severity: FluxSeverity.CAUTION)

        let result = fc.check(45)

        XCTAssertEqual(result.violatedLo, -30)
        XCTAssertEqual(result.violatedHi, 30)
    }

    func testPerformance() {
        let fc = FluxConstraint(sensorName: "performance")
        try! fc.fromPreset("aviation")

        self.measure {
            for i in 0..<10000 {
                _ = fc.check((i % 255) - 127)
            }
        }
    }
}

#endif

File: src/go/flux_constraint.go
package flux

import (
	"fmt"
	"math"
	"time"
)

// Severity levels for constraint violations
const (
	PASS     = 0
	CAUTION  = 1
	WARNING  = 2
	CRITICAL = 3
)

// FluxResult contains the result of a constraint check
type FluxResult struct {
	ErrorMask   uint8 // Bitmask of violated constraints (0-7)
	Severity    int   // Highest severity level encountered
	ViolatedLo  int8  // Lowest violated threshold
	ViolatedHi  int8  // Highest violated threshold
	CheckedValue int8 // The clamped input value that was checked
}

// FluxConstraint represents a single constraint with thresholds
type FluxConstraint struct {
	LoThreshold int8 // Lower threshold
	HiThreshold int8 // Upper threshold
	Severity    int  // Severity level for this constraint
}

// FluxChecker manages up to 8 constraints per sensor
type FluxChecker struct {
	Constraints [8]FluxConstraint
	NumActive   int // Number of active constraints
	SensorName  string
}

// NewFluxChecker creates a new constraint checker
func NewFluxChecker(sensorName string) *FluxChecker {
	return &FluxChecker{
		SensorName: sensorName,
		NumActive:  0,
	}
}

// AddConstraint adds a constraint (max 8 total)
func (fc *FluxChecker) AddConstraint(loThreshold, hiThreshold int8, severity int) error {
	if fc.NumActive >= 8 {
		return fmt.Errorf("maximum 8 constraints allowed")
	}
	if loThreshold > hiThreshold {
		return fmt.Errorf("low threshold cannot be greater than high threshold")
	}

	fc.Constraints[fc.NumActive] = FluxConstraint{
		LoThreshold: loThreshold,
		HiThreshold: hiThreshold,
		Severity:    severity,
	}
	fc.NumActive++
	return nil
}

// saturateINT8 clamps value to [-127, 127]
func saturateINT8(value int) int8 {
	if value > 127 {
		return 127
	}
	if value < -127 {
		return -127
	}
	return int8(value)
}

// Check performs constraint checking on a single value
func (fc *FluxChecker) Check(value int) FluxResult {
	clampedValue := saturateINT8(value)

	result := FluxResult{
		ErrorMask:    0,
		Severity:     PASS,
		ViolatedLo:   127,
		ViolatedHi:   -127,
		CheckedValue: clampedValue,
	}

	for i := 0; i < fc.NumActive; i++ {
		constraint := fc.Constraints[i]

		if clampedValue < constraint.LoThreshold || clampedValue > constraint.HiThreshold {
			// Constraint violated
			result.ErrorMask |= (1 << uint(i))

			if constraint.Severity > result.Severity {
				result.Severity = constraint.Severity
			}

			if constraint.LoThreshold < result.ViolatedLo {
				result.ViolatedLo = constraint.LoThreshold
			}

			if constraint.HiThreshold > result.ViolatedHi {
				result.ViolatedHi = constraint.HiThreshold
			}
		}
	}

	// Reset violated thresholds if no violations
	if result.ErrorMask == 0 {
		result.ViolatedLo = 0
		result.ViolatedHi = 0
	}

	return result
}

// BatchStats contains statistics from batch processing
type BatchStats struct {
	TotalChecks     int
	ViolationCount  int
	MaxSeverity     int
	AvgSeverity     float64
	ProcessingTime  time.Duration
}

// CheckBatch processes multiple values and returns results with statistics
func (fc *FluxChecker) CheckBatch(values []int) ([]FluxResult, BatchStats) {
	startTime := time.Now()

	results := make([]FluxResult, len(values))
	stats := BatchStats{
		TotalChecks: len(values),
		MaxSeverity: PASS,
	}

	totalSeverity := 0

	for i, value := range values {
		results[i] = fc.Check(value)

		if results[i].ErrorMask != 0 {
			stats.ViolationCount++
		}

		if results[i].Severity > stats.MaxSeverity {
			stats.MaxSeverity = results[i].Severity
		}

		totalSeverity += results[i].Severity
	}

	if len(values) > 0 {
		stats.AvgSeverity = float64(totalSeverity) / float64(len(values))
	}

	stats.ProcessingTime = time.Since(startTime)

	return results, stats
}

// FromPreset loads industry-standard constraint presets
func (fc *FluxChecker) FromPreset(presetName string) error {
	fc.NumActive = 0 // Reset constraints

	switch presetName {
	case "aviation":
		// Aviation sensor constraints (altitude, speed, engine parameters)
		fc.AddConstraint(-100, 100, PASS)      // Normal operation range
		fc.AddConstraint(-80, 80, CAUTION)     // Recommended range
		fc.AddConstraint(-60, 60, WARNING)     // Attention needed
		fc.AddConstraint(-40, 40, CRITICAL)    // Immediate action required

	case "medical":
		// Medical device constraints (vital signs, sensor readings)
		fc.AddConstraint(-50, 50, PASS)        // Normal physiological range
		fc.AddConstraint(-40, 40, CAUTION)     // Monitor closely
		fc.AddConstraint(-30, 30, WARNING)     // Clinical intervention
		fc.AddConstraint(-20, 20, CRITICAL)    // Emergency response

	case "maritime":
		// Maritime navigation constraints (compass, depth, weather)
		fc.AddConstraint(-120, 120, PASS)      // Full operational range
		fc.AddConstraint(-90, 90, CAUTION)     // Weather advisory
		fc.AddConstraint(-60, 60, WARNING)     // Navigation warning
		fc.AddConstraint(-30, 30, CRITICAL)    // Safety critical

	case "automotive":
		// Automotive sensor constraints (engine, safety systems)
		fc.AddConstraint(-110, 110, PASS)      // Normal driving range
		fc.AddConstraint(-85, 85, CAUTION)     // Performance monitoring
		fc.AddConstraint(-55, 55, WARNING)     // System alert
		fc.AddConstraint(-25, 25, CRITICAL)    // Safety shutdown

	case "energy":
		// Energy grid constraints (voltage, current, frequency)
		fc.AddConstraint(-127, 127, PASS)      // Full scale range
		fc.AddConstraint(-100, 100, CAUTION)   // Grid stability
		fc.AddConstraint(-75, 75, WARNING)     // Load shedding
		fc.AddConstraint(-50, 50, CRITICAL)    // Protection trip

	default:
		return fmt.Errorf("unknown preset: %s", presetName)
	}

	return nil
}

// Benchmark measures constraint checking performance
func (fc *FluxChecker) Benchmark(iterations int) (float64, error) {
	if iterations <= 0 {
		return 0, fmt.Errorf("iterations must be positive")
	}

	// Prepare test data
	testValues := make([]int, 1000)
	for i := range testValues {
		testValues[i] = (i % 255) - 127 // Range: -127 to 127
	}

	startTime := time.Now()

	checksPerformed := 0
	for i := 0; i < iterations; i++ {
		for _, value := range testValues {
			fc.Check(value)
			checksPerformed++
		}
	}

	duration := time.Since(startTime)
	checksPerSecond := float64(checksPerformed) / duration.Seconds()

	return checksPerSecond, nil
}

// Unit tests embedded in the same file
// To run: go test

//go:build test
// +build test

import (
	"testing"
)

func TestSaturateINT8(t *testing.T) {
	tests := []struct {
		input    int
		expected int8
	}{
		{0, 0},
		{127, 127},
		{128, 127},
		{-127, -127},
		{-128, -127},
		{1000, 127},
		{-1000, -127},
	}

	for _, test := range tests {
		result := saturateINT8(test.input)
		if result != test.expected {
			t.Errorf("saturateINT8(%d) = %d, expected %d", test.input, result, test.expected)
		}
	}
}

func TestAddConstraint(t *testing.T) {
	fc := NewFluxChecker("test")

	// Test normal addition
	err := fc.AddConstraint(-50, 50, WARNING)
	if err != nil {
		t.Errorf("Failed to add constraint: %v", err)
	}

	// Test invalid threshold order
	err = fc.AddConstraint(50, -50, WARNING)
	if err == nil {
		t.Error("Expected error for invalid threshold order")
	}

	// Test maximum constraints
	for i := 1; i < 8; i++ {
		fc.AddConstraint(-10, 10, PASS)
	}

	err = fc.AddConstraint(-5, 5, PASS)
	if err == nil {
		t.Error("Expected error when adding 9th constraint")
	}
}

func TestBasicCheck(t *testing.T) {
	fc := NewFluxChecker("test")
	fc.AddConstraint(-50, 50, WARNING)

	// Test within bounds
	result := fc.Check(25)
	if result.ErrorMask != 0 {
		t.Error("Expected no violations for value within bounds")
	}

	// Test out of bounds
	result = fc.Check(75)
	if result.ErrorMask == 0 {
		t.Error("Expected violation for value out of bounds")
	}

	if result.Severity != WARNING {
		t.Errorf("Expected severity %d, got %d", WARNING, result.Severity)
	}
}

func TestMultipleConstraints(t *testing.T) {
	fc := NewFluxChecker("test")
	fc.AddConstraint(-100, 100, PASS)
	fc.AddConstraint(-50, 50, CAUTION)
	fc.AddConstraint(-25, 25, WARNING)
	fc.AddConstraint(-10, 10, CRITICAL)

	result := fc.Check(75)

	// Should violate constraints 1, 2, and 3 (bits 1, 2, 3 set)
	expectedMask := uint8(0b00001110)
	if result.ErrorMask != expectedMask {
		t.Errorf("Expected error mask %08b, got %08b", expectedMask, result.ErrorMask)
	}

	if result.Severity != CRITICAL {
		t.Errorf("Expected CRITICAL severity, got %d", result.Severity)
	}
}

func TestCheckBatch(t *testing.T) {
	fc := NewFluxChecker("test")
	fc.AddConstraint(-50, 50, WARNING)

	values := []int{-25, 0, 25, 75, -75}
	results, stats := fc.CheckBatch(values)

	if len(results) != len(values) {
		t.Errorf("Expected %d results, got %d", len(values), len(results))
	}

	if stats.TotalChecks != len(values) {
		t.Errorf("Expected %d total checks, got %d", len(values), stats.TotalChecks)
	}

	if stats.ViolationCount != 2 {
		t.Errorf("Expected 2 violations, got %d", stats.ViolationCount)
	}
}

func TestFromPreset(t *testing.T) {
	fc := NewFluxChecker("test")

	presets := []string{"aviation", "medical", "maritime", "automotive", "energy"}

	for _, preset := range presets {
		err := fc.FromPreset(preset)
		if err != nil {
			t.Errorf("Failed to load preset %s: %v", preset, err)
		}

		if fc.NumActive == 0 {
			t.Errorf("Preset %s loaded no constraints", preset)
		}
	}

	// Test invalid preset
	err := fc.FromPreset("invalid")
	if err == nil {
		t.Error("Expected error for invalid preset")
	}
}

func TestBenchmark(t *testing.T) {
	fc := NewFluxChecker("test")
	fc.FromPreset("aviation")

	rate, err := fc.Benchmark(10)
	if err != nil {
		t.Errorf("Benchmark failed: %v", err)
	}

	if rate <= 0 {
		t.Errorf("Expected positive benchmark rate, got %f", rate)
	}

	// Test invalid iterations
	_, err = fc.Benchmark(0)
	if err == nil {
		t.Error("Expected error for zero iterations")
	}
}

func TestSaturationBehavior(t *testing.T) {
	fc := NewFluxChecker("test")
	fc.AddConstraint(-50, 50, WARNING)

	// Test extreme values get saturated
	result := fc.Check(1000)
	if result.CheckedValue != 127 {
		t.Errorf("Expected saturated value 127, got %d", result.CheckedValue)
	}

	result = fc.Check(-1000)
	if result.CheckedValue != -127 {
		t.Errorf("Expected saturated value -127, got %d", result.CheckedValue)
	}
}

func TestViolatedThresholds(t *testing.T) {
	fc := NewFluxChecker("test")
	fc.AddConstraint(-30, 30, WARNING)
	fc.AddConstraint(-60, 60, CAUTION)

	result := fc.Check(45)

	if result.ViolatedLo != -30 {
		t.Errorf("Expected violated_lo -30, got %d", result.ViolatedLo)
	}

	if result.ViolatedHi != 30 {
		t.Errorf("Expected violated_hi 30, got %d", result.ViolatedHi)
	}
}

// Benchmark function for go test -bench
func BenchmarkFluxCheck(b *testing.B) {
	fc := NewFluxChecker("benchmark")
	fc.FromPreset("aviation")

	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		fc.Check((i % 255) - 127)
	}
}